### PR TITLE
ENH: stats: add `keepdims` support to `_axis_nan_policy` decorator

### PR DESCRIPTION
**Pull request by tirthasheshpatel**
#### Reference issue

gh-14651

Discussed in gh-15472 and gh-15423 

#### What does this implement/fix?

Some stats functions accept the `keepdims` parameter which makes it difficult to apply the `_axis_nan_policy_factory` decorator. This PR adds the `keepdims` parameter to `_axis_nan_policy_factory` so all the functions that use the decorator get it. NumPy is quite consistent about adding this to its reduction functions so it seems like a reasonable enhancement to add this to the stats functions as well. I also added a basic test that checks if the output has the correct shape and is consistent with the output of `keepdims=False`. The test runs for all the functions using the decorator.

Once this is merged, it would make it easy to add the decorator to `variation`, `mode`, and `iqr`.

#### Additional information

@mdhaber Let me know if this needs more tests.

@tupui @Kai-Striega As usual, would appreciate if you could play with this a little and provide your feedback! :)

### CODE REVIEW

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:243]: This needs the same treatment for tuple axes as above. How about a separate `_expand_dims` function?

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:237]: I'd expect this could run into trouble if `reduced_axes` is not sorted. I'm not certain there is a problem, but please consider it. If there is a problem, the tests wouldn't catch it because they are all sorted in an order this works for. The `(1, -1)` test comes closest to catching it, but doesn't because `-1` is equivalent to `3`.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:494]: It's unfortunate that there are already so many return statements. Consequently, I think you missed (at least) one - original line 435, new line 477.

I tried to limit separate `return`s to fundamentally different cases that would be very awkward to combine, but maybe I could have done better. I'd suggest that philosophy in this PR, though. Can this be restructured so that this `return` turns into a `res = `?

Consider whether you'd prefer to pass `keepdims` as an argument into the function and put the `if` statement into the function rather than always prefacing `_add_reduced_axes` with `if keepdims`.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:400]:
```suggestion
                keepdims = kwds.pop("keepdims", False)
```
Doesn't this work?
Also, I think this belongs below, where we do `kwds.pop('axis'...)` and `kwds.pop('nan_policy'...)`.

Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:400]:
```suggestion
```
See below.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:401]:
```suggestion
```

**Comment by mdhaber** [comment]
We should probably send an email to the mailing list after approval but before merge.

File: scipy/stats/_axis_nan_policy.py
Comment (tirthasheshpatel) [reply scipy/stats/_axis_nan_policy.py]: Done.

File: scipy/stats/_axis_nan_policy.py
Comment (tirthasheshpatel) [reply scipy/stats/_axis_nan_policy.py]: > I'd expect this could run into trouble if `reduced_axes` is not sorted.

You are right. Thanks for catching this. `_broadcast_shapes` function already sorted and validated axis. I have separated that part into a `_process_axis` function and used it sort `reduced_axis`.

One catch is that it is possible that the user didn't pass any samples (which leads to an error down the line), in which case, `samples` array is empty and processing axis doesn't make any sense in that case. To handle this, I have taken as input the shapes of the samples and, if it is empty, we just return the axis as is. I thought of passing the `n_dims` parameter instead, but if there are no samples, there is no way to calculate it (alternative is to pass a invalid value like `-1` to indicate failure).

File: scipy/stats/_axis_nan_policy.py
Comment (tirthasheshpatel) [reply scipy/stats/_axis_nan_policy.py]: > I think you missed (at least) one - original line 435, new line 477

That return statement just returns a scalar `nan` for each output. As the output there isn't consistent with the shape of other outputs, I wasn't sure how would `keepdims` behave there. I thought it's best to keep it untouched but this needs to be discussed further.

File: scipy/stats/_axis_nan_policy.py
Comment (tirthasheshpatel) [reply scipy/stats/_axis_nan_policy.py]: > Doesn't this work?

Oh yes, forgot `pop` takes another input as default return. I have changed it this way and moved it with other pops.

File: scipy/stats/_axis_nan_policy.py
Comment (tirthasheshpatel) [reply scipy/stats/_axis_nan_policy.py]: > Consider whether you'd prefer to pass `keepdims` as an argument into the function and put the `if` statement into the function rather than always prefacing `_add_reduced_axes` with `if keepdims`

I tried this but `result_unpacker` confuses me. I thought it is supposed to unpack the `result_object` but turns out it is also used to unpack the result returned by `np.apply_along_axis` (see last return). So, I was hesitant to apply `result_unpacker` everywhere before passing it to the `_add_reduced_axes`.

> Can this be restructured so that these new `return`s turn into `res = `?

Done.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:458]: Hmm we write that the output array will be broadcastable with the inputs, but if we use `max` here, that's not true. Maybe we want `min`, actually?

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:587]: This is still a return.

Comment (tirthasheshpatel) [reply scipy/stats/_axis_nan_policy.py:587]: When `keepdims` is True, the `result_unpacker` is applied (to get a tuple of results) and then the `result_object` before returning from `_add_reduced_axes`. The default `result_unpacker` doesn't work when the result is an instance of the `result_object` (so, the line below this one `return result_object(*result_unpacker(res))` fails). See https://github.com/scipy/scipy/pull/15478#discussion_r795037172

This is why it is tricky to make it work without the `if keepdims` before it and returning from there. I think we can avoid applying the result_object in the `_add_reduced_axes` function but I thought it might make things a little more confusing and unreadable.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:34]: This diff turned out really confusing : / Can it be moved so GitHub doesn't get so confused?

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:38]: Just a thought: if we don't like special cases for `axis` `None`, we could turn that into `axis = np.arange(n_dims)` here.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:42]: You mentioned `np.AxisError`. This would be a fine time to switch it.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [reply scipy/stats/_axis_nan_policy.py]: > That return statement just returns a scalar nan for each output. As the output there isn't consistent with the shape of other outputs, I wasn't sure how would keepdims behave there. 

That part of the decorator starting between `# if axis is not needed, just handle nan_policy and return` and `# check for empty input` is for when all samples are 1D - whether that's because they have been revealed due to the value of `axis` or whether they were passed in as 1D. So yes, currently it returns a scalar `nan` for each output, but that's the correct behavior. It's not inconsistent, because a few lines later, `return hypotest_fun_out(*samples, **kwds)` will return a scalar value for each output because we're still in the part of the decorator where all the samples are 1D. 

Both need to change if we're adding back dimensions.

> this needs to be discussed further.

It definitely does when we apply the decorator to `mode` because currently this assumes that the correct behavior for `propagate` is to return a NaN. That's not true for `mode`. We'll want to add a parameter to the decorator factory like, e.g. `nan_propagate_override=True`. When True (default and current behavior), the decorator doesn't even execute the function if there are NaNs in the slice. When set False, the decorator actually executes the function as usual. I think this will mean that the line:
```python3
if any(contains_nans) and nan_policy == 'propagate':
```
becomes
```python3
if any(contains_nans) and nan_policy == 'propagate' and nan_propagate_override:
```
in the 1D section and similarly in the ND section.

Comment (mdhaber) [reply scipy/stats/_axis_nan_policy.py]: > I tried this but result_unpacker confuses me. I thought it is supposed to unpack the result_object but turns out it is also used to unpack the result returned by np.apply_along_axis (see last return).

The idea is that it's written either way. `result_unpacker` ensures that each separate output of the function is an element of a sequence (tuple). It becomes especially important when the object returned by function is not a sequence / cannot be unpacked like a tuple. 

See, for instance, commit f27d887d7107405b613426a5e8301e8476ca6f5b in which I removed the decorator from a bunch of functions I tested it on during development. `CramerVonMisesResults` are not iterable. They are not namedtuples like most stats function returns; they just are objects with fields `statistic` and `value`. The function isn't vectorized, either, so the decorator vectorizes it using `np.vectorize`.  Suppose we call `cramervonmises` on an `M` x `N` input with `axis=0`. `np.vectorize` produces an object array of shape `(N,)` in which each element is a `CramerVonMisesResult`. Before we return it to the user, we need to unpack it and repack it into a single `CramerVonMisesResult` in which `statistic` and `pvalue` are both numerical arrays of shape `(N,)`:
```
result_unpacker=np.vectorize(lambda res: (res.statistic, res.pvalue))
```
This use of `np.vectorize` loops over the object array to pick out the `statistic` and `pvalue`, and the result is a tuple of numerical arrays of shape `(N,)`. Maybe this will help:
```python3
from dataclasses import make_dataclass
Result = make_dataclass('Result', ('statistic', 'pvalue'))
res = np.array([Result(1, 2), Result(3, 4), Result(5, 6)])
res  # array([Result(statistic=1, pvalue=2), Result(statistic=3, pvalue=4), Result(statistic=5, pvalue=6)], dtype=object)
result_unpacker=np.vectorize(lambda res: (res.statistic, res.pvalue))
res2 = result_unpacker(res)
res2  # (array([1, 3, 5]), array([2, 4, 6]))
```
This works to unpack a single `Result` object, too.
```python3
res3 = result_unpacker(Result(1, 2))
res3  # (array(1), array(2))
```
Anyway, hope that helps to understand what it's for. We should write the new code the way we think it should be, without having to tiptoe around what's there.

File: scipy/stats/_axis_nan_policy.py
Comment (tirthasheshpatel) [reply scipy/stats/_axis_nan_policy.py:34]: > This diff turned out really confusing : /

Yeah, there was a lot going on behind the scenes than I thought I would need to know to make this work. Sorry! I will go over this again and simplify it as much as possible.



> I see you factored this out so it could be used below. Otherwise, is added/changed that is different from what was going on in `broadcast_shapes` before?

No, I just factored out the axis validation and sorting part. Nothing changed.

Comment (mdhaber) [reply scipy/stats/_axis_nan_policy.py:34]: No, it's not a problem with the code. It's just that GitHub got confused. I wouldn't refactor this expect consider moving the function so that GitHub doesn't mix up the two functions. We can move it back here later.

File: scipy/stats/_axis_nan_policy.py
Comment (tirthasheshpatel) [reply scipy/stats/_axis_nan_policy.py]: > Anyway, hope that helps to understand what it's for.

Thanks, this was a great explanation! I think I got the purpose of `result_unpacker` and what it's doing.

> We should write the new code the way we think it should be, without having to tiptoe around what's there.

Makes sense. I will rethink this and hopefully simplify all the tiptoeing!

> That part of the decorator starting between `# if axis is not needed, just handle nan_policy and return` and `# check for empty input` is for when all samples are 1D - whether that's because they have been reveled due to the value of `axis` or whether they were passed in as 1D.

Oops, I forgot thers's `if np.all(ndims <= 1)` condition! Will change this and add tests for it.

File: scipy/stats/_axis_nan_policy.py
Comment (tirthasheshpatel) [reply scipy/stats/_axis_nan_policy.py:42]: Will change it here.

File: scipy/stats/_axis_nan_policy.py
Comment (tirthasheshpatel) [reply scipy/stats/_axis_nan_policy.py]: Good catch, this got me thinking! I think both should work: since we are anyways going to return an array of shape `(1,1,1,...)`, if all other arrays broadcast with each other, they must also broadcast with an array of all 1s. Even if we return array with `min` ndims, it would be filled with 1s on the right when broadcasting. Does this sound right or did you have any case in mind where `min` is desirable over `max`?

File: scipy/stats/_axis_nan_policy.py
Comment (tirthasheshpatel) [reply scipy/stats/_axis_nan_policy.py]: Removed `if keepdims` from everywhere and use `_add_reduced_axes` to remaining `return` statements. Tests for them also added.

File: scipy/stats/_axis_nan_policy.py
Comment (tirthasheshpatel) [reply scipy/stats/_axis_nan_policy.py]: Merged the two return statements in [latest commits](https://github.com/scipy/scipy/pull/15478/commits/54f0b09bc953a3deadaa56245cee4007e7ab66e5).

File: scipy/stats/_axis_nan_policy.py
Comment (tirthasheshpatel) [reply scipy/stats/_axis_nan_policy.py]: Reverted for now. Will add this change back before merging.

File: scipy/stats/_axis_nan_policy.py
Comment (tirthasheshpatel) [reply scipy/stats/_axis_nan_policy.py]: It would! I tried it locally but it gave me some non-trivial errors. Maybe, tackle that in another PR?

File: scipy/stats/_axis_nan_policy.py
Comment (tirthasheshpatel) [reply scipy/stats/_axis_nan_policy.py]: Done

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [reply scipy/stats/_axis_nan_policy.py]: OK. Let's remember to come back to this if you think there is something wrong.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [reply scipy/stats/_axis_nan_policy.py]: OK. I was going to suggest that you could just move `_process_axes` somewhere else - after `_broadcast_shapes`, for instance - so that GitHub didn't think you were heavily modifying `_broadcast_shapes`. We want it to see `_process_axes` as a totally new function, and we want it to see that some of `_broadcast_shapes` has been deleted and replaced with a call to `_process_axes`. We can manually check that `_process_axes` basically just contains all the lines deleted from `_broadcast_shapes`. Does that make sense?

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [reply scipy/stats/_axis_nan_policy.py]: I see what you're saying.
You are right that all of the input and output arrays will broadcast together either way.
I guess my thought was that if you use `min` instead of `max`, then the _output_ array will broadcast _to_ the shape of any of the input arrays individually.
On the other hand, if you use `max`, then even though the input and output arrays are "broadcastable", the output array will not necessarily broadcast to the shape of each of the input arrays.
Does that make sense?
Here's an example:
```python3
x = np.random.rand(3, 4)
y = np.random.rand(5, 3, 1)

out1 = (x*y).sum(axis=None).reshape((1, 1))  # `min`
np.broadcast(x, y, out1)   # no problem
np.broadcast_to(out1, x.shape)   # no problem
np.broadcast_to(out1, y.shape)   # no problem

out1 = (x*y).sum(axis=None).reshape((1, 1, 1))  # `max`
np.broadcast(x, y, out1)   # no problem
np.broadcast_to(out1, x.shape)   # problem!
np.broadcast_to(out1, y.shape)   # no problem
```
That said, I think either way satisfies the statement:
> With this option, the result will broadcast correctly against the input array.

because "against" doesn't imply a direction very strongly. 

My thought is just that `min` is slightly more flexible than `max`.

On the other hand, the argument name `keepdims` suggests that we keep all the dimensions we have before the statistic is taken. 
```out1 = (x*y).sum(axis=None, keepdims=True).shape  # `(1, 1, 1)`
```
So that's an argument in favor of `max`.

Can you find any two-"sample" NumPy functions we can use as precedent? (This only matters when there are multiple "samples" that might be different shapes.) I tried for a bit without luck.

**Comment by mdhaber** [comment]
Getting really close. I still think we can simplify a bit to get rid of special cases (e.g. `if n_samples == 1`). I think it's worth the extra effort now to keep this (very complicated) function easier to maintain. I'll give it a shot shortly.

File: scipy/stats/_axis_nan_policy.py
Comment (tirthasheshpatel) [reply scipy/stats/_axis_nan_policy.py]: Oh, OK. Moved the _procees_axes function below _broadcast_shapes. The diff is better now.

File: scipy/stats/_axis_nan_policy.py
Comment (tirthasheshpatel) [reply scipy/stats/_axis_nan_policy.py]: > Does that make sense?

Makes sense but I have one more argument in favor of `max` :)

> On the other hand, the argument name `keepdims` suggests that we keep all the dimensions we have before the statistic is taken.

To me, `max` makes intuitive sense because of this argument. Also, with this, the behavior of `axis=None` and `axis=(0, 1, 2, ...)` matches; both fill in max number of 1s. For example:

```py
>>> import numpy as np
>>> from scipy import stats
>>> x = np.random.rand(3, 4, 5)
>>> y = np.random.rand(4, 5)
>>> res = stats.kruskal(x, y, axis=None, keepdims=True)
>>> res[0].shape
(1, 1, 1)
>>> res = stats.kruskal(x, y, axis=(0, 1, 2), keepdims=True)
>>> res[0].shape
(1, 1, 1)  # axis=(0, 1, 2) is valid and returns array with `max` dims of 1s.
```

In the above example, `axis=(0, 1, 2)` is the longest valid axis tuple. If we assume `axis=None` to be equivalent to it, `max` rather than `min` behaves correctly.

> Can you find any two-"sample" NumPy functions we can use as precedent?

I too can't find any numpy function that takes more than one sample with different shapes. If you agree to the above argument, let's keep `max` otherwise I will change it to `min`. I am OK with both.

File: scipy/stats/_axis_nan_policy.py
Comment (tupui) [inline scipy/stats/_axis_nan_policy.py:245]: Actually (this is on main) SciPy 1.9 will only supports Python 3.8 hence NumPy >=1.18.5 😃 So we can already remove this.

Comment (tirthasheshpatel) [reply scipy/stats/_axis_nan_policy.py:245]: That's great! Will update this.

Comment (tupui) [reply scipy/stats/_axis_nan_policy.py]: Yeah it seems to happen a lot recently. Does not mean it's this PR.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [reply scipy/stats/_axis_nan_policy.py]: Yup let's leave it as `max`. Good to make the decision explicitly, though!

**Comment by mdhaber** [comment]
Didn't make it to this today : / In the meantime, @tirthasheshpatel, could you take a look at gh-15423 and merge if you're happy with it? It already has two approvals.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:246]: We need to work on the rest of this until this function is just:
```suggestion
    res = ([np.expand_dims(output, reduced_axes) for output in res] if keepdims
           else res)
    return result_object(*res)
```
And we don't want to have to do things like 
```
            if n_outputs > 1:
                res = result_unpacker(res)
```
In some places but not others. If we do it in one place, we should do it everywhere. I think we can get this so we always apply the `result_unpacker`.

With some slight tweaks to other parts, I'm sure we can get there. 

For the ndarray part of the decorator, the reason it's tricky is that `result_unpacker` wasn't designed to work with the raw output of a function when we use the function's native vectorization. It was only designed to work with output of a function that _we_ vectorized using `np.apply_along_axis`. Roughly speaking, for multi-output functions: 
- when we use the native vectorization, the different outputs come along the zeroth axis (because the output is really a tuple of arrays)
- when we vectorize using `np.apply_along_axis`, the different outputs come along the _last_ axis because I standardized the statistic to be taken along the last axis. The fix is probably to standardize to take the statistic along the zeroth axis. I ran into this same sort of thing recently in gh-15491.

Give me a few more days to play with this.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:475]:
```suggestion
```

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:492]: We need to ensure that `res` is a sequence of outputs.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:518]: Similarly here, `res` needs to be a sequence of outputs.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:561]:
```suggestion
            res = result_unpacker(res)
```
This ensures that `res` is a sequence of outputs

File: scipy/stats/_axis_nan_policy.py
Comment (tirthasheshpatel) [reply scipy/stats/_axis_nan_policy.py:246]: > * When we vectorize using `np.apply_along_axis`, the different outputs come along the _last_ axis because I standardized the statistic to be taken along the last axis. The fix is probably to standardize to take the statistic along the zeroth axis.

That sounds good. I was hesitant to touch that part since I didn't know the reason it was there.

> Give me a few more days to play with this.

Let me know if you want me to make the changes. I think I understand the issue and can get it there.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:559]:
```suggestion
            x = np.moveaxis(x, axis, 0)
            res = np.apply_along_axis(hypotest_fun, axis=0, arr=x)
```

**Comment by mdhaber** [comment]
@tirthasheshpatel I see you already merged tirthasheshpatel/scipy#12 - does it make sense? I added a few comments about the reasoning there that might be helpful.
I also suggested that we could rename `result_unpacker` to `result_tuplizer` if you think it would be more descriptive.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:233]:
```suggestion
    if keepdims = True. `res` must be a sequence results.
```

**Comment by tirthasheshpatel** [comment]
> Or maybe `result_unpacker` -> `result_to_tuple` and `result_object` -> `tuple_to_result` to make the inverse nature of the two very obvious.

Sounds much better to me. This would make things less confusing (at least for me). I will make the change in the next commit.

File: scipy/stats/tests/test_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/tests/test_axis_nan_policy.py:355]: This test might be fine (I haven't read carefully yet). I like that it checks something about `axis` `None`, negative, a tuple, and an out of order tuple. Seems like all the major cases.

On one hand, I say I like to write tests that suggest correct functionality for all possible inputs without knowledge of how the code works. This leads to very exhaustive tests. I tried to do that in the initial `_axis_nan_policy_test`. But as we expand the decorator, I don't think it's necessary. That's why we wrote separate, simpler tests when we added `axis` tuple and masked array support to the decorator rather than adding an additional layer of `parametrize`ation to the `test_axis_nan_policy_full`/`test_axis_nan_policy_fast`. It would be ok with me if you wanted to do that here rather than testing for all hypotests. I think we just need to hit these major points:
* single-output and multi-output functions (e.g. a function like `gmean` and a function like `mannwhitneyu`)
* Axis negative, positive, `None`, and tuple (as you have above)
* 1D with no NaNs, 1D with NaN propagation, zero size output, natively vectorized function with multidimensional output, and function vectorized with the decorator (either because it doesn't have native vectorization or because we're handling the vectorization because the input has NaNs or masked values)

Does that make sense? Do we have those cases here? Are there others you think are important?

Comment (tirthasheshpatel) [reply scipy/stats/tests/test_axis_nan_policy.py:355]: > Does that make sense? Do we have those cases here? Are there others you think are important?

You are right. On my machine, `test_axis_nan_policy` already takes a minute to run (1/3 the time taken to test the entire module). I will remove parameterization and we should probably start removing some `test_mstats` tests that overlap with that of `test_axis_nan_policy` and also the code from stats functions that call the mstats module. (I don't think we need those calls anymore)

Comment (mdhaber) [reply scipy/stats/tests/test_axis_nan_policy.py:355]: It takes <4s to run `test_mstats_basic.py` and `test_extras.py`. I'd prefer to just remove the test files if and when we remove `mstats` altogether.

Comment (tirthasheshpatel) [reply scipy/stats/tests/test_axis_nan_policy.py]: I have pushed a test that only tests `gmean` and `mannwhitneyu` with all the cases you outlined in your comment.

> * natively vectorized function with multidimensional output, and function vectorized with the decorator

I am sorry, I don't follow this. How should I natively vectorize since hypotests aren't elementwise operations?

Comment (mdhaber) [reply scipy/stats/tests/test_axis_nan_policy.py]: No, that's not what I meant. There were five items in that single bullet:

- 1D with no NaNs,
- 1D with NaN propagation
- zero size output
- natively vectorized function with multidimensional output
- and function vectorized with the decorator (either because it doesn't have native vectorization or because we're handling the vectorization because the input has NaNs or masked values)

Each of these items corresponds with one of five important return statements in the decorator. I wanted to make sure all of these relevant return statements (and the code paths to get to them) are tested.

File: scipy/stats/_axis_nan_policy.py
Comment (mdhaber) [inline scipy/stats/_axis_nan_policy.py:304]: This should be reverted. Look at the calls to `tuple_to_result` - the individual components need to be provided (but they are always coming from a tuple, which is why this is not an unreasonable name).
```suggestion
        statistical test) provided the separate components (e.g. statistic,
        pvalue).
```
The reason a `tuple_to_result` callable should accept separate arguments rather than a single tuple is that the typical `tuple_to_result` is the initializer of a hypothesis test result class or namedtuple.

**Review by mdhaber — APPROVED** [review]
Aside from the comment, I think this is ready. The only comment from the last review was to think about condensing the tests, and it looks like you've done that.

Please send an email to the mailing list about adding `keepdims` to `scipy.stats` reduction operations to make our interface more internally consistent and more consistent with the interface of NumPy reduction operations.  We'll give it a week for feedback, then merge if there are no more comments.

**Comment by tirthasheshpatel** [comment]
[Mail sent](https://mail.python.org/archives/list/scipy-dev@python.org/thread/JZR4U5N23YWGKUUGZQYCVTEMUI5GW6TN/).

**Comment by mdhaber** [comment]
Thanks @tirthasheshpatel!

**Comment by tirthasheshpatel** [comment]
Thanks for the reviews @mdhaber!

**Pull request merged**
